pyg_lib/csrc/sampler/cuda/subgraph_kernel.cu

```diff
@@ -0,0 +1,90 @@
+#include <ATen/ATen.h>
+#include <torch/library.h>
+
+#include "pyg_lib/csrc/utils/cuda/helpers.h"
+
+namespace pyg {
+namespace sampler {
+
+namespace {
+
+#define FULL_MASK 0xffffffff
+
+template <typename scalar_t>
+__global__ void subgraph_deg_kernel_impl(
+    const scalar_t* __restrict__ rowptr_data,
+    const scalar_t* __restrict__ col_data,
+    const scalar_t* __restrict__ nodes_data,
+    const scalar_t* __restrict__ to_local_node_data,
+    scalar_t* __restrict__ out_data,
+    int64_t num_nodes) {
+  CUDA_1D_KERNEL_LOOP(scalar_t, thread_idx, 32 * num_nodes) {
+    scalar_t i = thread_idx >> 5;           // thread_idx / 32
+    scalar_t lane = thread_idx & (32 - 1);  // thread_idx % 32
+
+    auto v = nodes_data[i];
+
+    scalar_t deg = 0;
+    for (scalar_t j = rowptr_data[v] + lane; j < rowptr_data[v + 1]; j += 32) {
+      if (to_local_node_data[col_data[j]] >= 0)  // contiguous access
+        deg++;
+    }
+
+    for (scalar_t offset = 16; offset > 0; offset /= 2)  // warp-level reduction
+      deg += __shfl_down_sync(FULL_MASK, deg, offset);
+
+    if (lane == 0)
+      out_data[i] = deg;
+  }
+}
+
+std::tuple<at::Tensor, at::Tensor, c10::optional<at::Tensor>> subgraph_kernel(
+    const at::Tensor& rowptr,
+    const at::Tensor& col,
+    const at::Tensor& nodes,
+    const bool return_edge_id) {
+  TORCH_CHECK(rowptr.is_cuda(), "'rowptr' must be a CUDA tensor");
+  TORCH_CHECK(col.is_cuda(), "'col' must be a CUDA tensor");
+  TORCH_CHECK(nodes.is_cuda(), "'nodes' must be a CUDA tensor");
+
+  const auto stream = at::cuda::getCurrentCUDAStream();
+
+  // We maintain a O(N) vector to map global node indices to local ones.
+  // TODO Can we do this without O(N) storage requirement?
+  const auto to_local_node = nodes.new_full({rowptr.size(0) - 1}, -1);
```

Comment:
Does N means the number of nodes in the graph?
What if we could filter each node in `nodes_data` since it should be much smaller than `rowptr_data`.
Otherwise we may consider caching this tensor to reduce memory allocation for each time.

Reply:
Good points! We use this vector as the inverse mapping from global node indices to new local ones. In C++, we use a map for this but can't do the same in CUDA. I will ping NVIDIA if they have an elegant solution for this.

Caching is an option as well, but requires a (non-intuitive and backend-specific) change in input arguments. I added it as a TODO for now.

Reply:
There's GPU hash table/set which may require some atomic operations when you build it, but lookup is fast.
I found that caching is not a good option since you have to reset the array every time.

Since you can sample on GPU, then the graph is not that big, a node array is not that bad and can make the code less complicated